publish-build.sh

```diff
@@ -17,51 +17,68 @@ set -x # print all commands
 shopt -s nullglob globstar
 # end: shellharden
 
+function printerr () {
+    >&2 echo -e "\e[91m${1}\e[0m"
+}
+
 if [[ ! ${TRAVIS_TAG:-} ]]; then
-    echo "Not a built for a Git tag, do not publish!"
+    printerr "Not built from a Git tag, do not publish!"
     exit 0
 fi
 
+DRYRUN=0
+if [ $# -gt 0 ] && [ "$1" == "--dry-run" ]; then
+    DRYRUN=1
+    printerr "PERFORMING DRY RUN"
+fi
+
 # this part needs to be expanded for other dist-tags
 # based on api levels
 
-function publishPackage () {
-    local COMPONENT=$1
-    local REPO_DIR=$2
-
-    COMPONENT=${COMPONENT//_/-}
+function exec () {
+    local CMD=$1
 
-    BUILD_REPO_NAME="${COMPONENT}"
-    BUILD_REPO_DIR="tmp/${BUILD_REPO_NAME}"
-
-    version=$(node -pe "require('./package.json').version")
-    echo "Publishing version: ${version}"
+    if [ $DRYRUN -eq 1 ]; then
+        echo $CMD
+    else
+        `${CMD}`
+    fi
+}
 
+function publishPackage () {
+    local PACKAGE_DIR=$1
+    local PACKAGE_JSON="${PACKAGE_DIR}package.json"
 
-    if [[ -d "$BUILD_REPO_DIR" ]]; then
-        npm publish "$BUILD_REPO_DIR" --tag "$DIST_TAG" --access public
+    if [[ ! -e ${PACKAGE_JSON} ]]; then
+        printerr "Package.json file '${PACKAGE_JSON}' does not exist, skipping publish."
     else
-        npm publish --tag "$DIST_TAG" --access public
+        name=$(node -pe "require('${PACKAGE_DIR}package.json').name")
+        version=$(node -pe "require('${PACKAGE_DIR}package.json').version")
+        echo "Publishing package: ${name} @ ${version}"
+
+        exec "npm publish \"$PACKAGE_DIR\" --tag \"$DIST_TAG\" --access public"
     fi
 }
 
 DIST_TAG=latest
-
-ROOT=$(basename $(pwd))
+BUILDS_DIR="./tmp/builds"
 
 echo "//registry.npmjs.org/:_authToken=${NPM_TOKEN}" > ~/.npmrc
 echo "//registry.npmjs.org/:username=travis4dhis2" >> ~/.npmrc
 echo "//registry.npmjs.org/:email=deployment@dhis2.org" >> ~/.npmrc
 
-if [[ ! -d "packages" ]]; then
+if [ ! -d "./packages" ] && [ ! -d "${BUILDS_DIR}" ]; then
     dir=$(pwd)
-    publishPackage "$ROOT" "$dir"
+    publishPackage "${dir}/"
```

Comment:
I think this is inconsistent between L72, L76, and L81. There is a trailing `/` at L72, but it is omitted at L76 and L81. And then, there is an assumption of the trailing slash at L50.

My personal preference is to add the slash when the interpolation is done rather than assuming it is there:

```
50    local PACKAGE_JSON="${PACKAGE_DIR}package.json"
++    local PACKAGE_JSON="${PACKAGE_DIR}/package.json" <-- i prefer this
```

Reply:
I prefer that way too, didn't want to mess too much with the for loops though.  In my testing, `pwd` resolved without a trailing slash but `$dir` always resolved with a trailing slash in `for dir in ./packages/*/`, hence the inconsistency between L72, L76, and L81.  I think it might be the final slash in that globstar pattern, will see if it works to do `./packages/*` and then insert the intervening slash at L50.

Reply:
As mentioned on slack, I overlooked the detail with the `packages/*/` glob, which only matches directories, whereas `packages/*` would give all files as well.

My preferences remain the same, but a clarification on the discrepancy between the three calls would be good for at-a-glance understanding.

Reply:
Uploaded to trim the trailing / in the for loop case, should be easier to follow now